spec/Section 6 -- Execution.md

```diff
@@ -565,6 +555,30 @@ DoesFragmentTypeApply(objectType, fragmentType):
 Note: The steps in {CollectFields()} evaluating the `@skip` and `@include`
 directives may be applied in either order since they apply commutatively.
 
+### Root Field Collection
+
+Root field collection processes the operation's top-level selection set:
+
+CollectRootFields(rootType, operationSelectionSet, variableValues):
+
+- Initialize {groupedFieldSet} to an empty ordered map of lists.
+- Call {CollectFields(rootType, operationSelectionSet, groupedFieldSet)}.
+- Return {groupedFieldSet}.
+
+### Object Subfield Collection
+
+Root field collection processes the operation's top-level selection set:
+
+CollectSubfields(objectType, fields, variableValues):
+
+- Initialize {groupedSubfieldSet} to an empty ordered map of lists.
+- For each {field} in {fields}:
+  - Let {fieldSelectionSet} be the selection set of {field}.
+  - If {fieldSelectionSet} is null or empty, continue to the next field.
+  - Call {CollectFields(objectType, fieldSelectionSet, variableValues,
+    groupedSubfieldSet)}.
```

Comment:
I don't love that we are storing state in a mutable reference passed in from above. I get that may be how graphql-js implements it, but IMO it would be better if `CollectFields` *returned* `groupedFieldSet`, and this algorithm *merged* those returned fields. Or CollectFields returns passed-in-set + fields-visited-set

Reply:
works for me! 0ce2f8e